묵찌빠 프로그램 [STEP 1] Jin, easy

wlsdud-dev opened the pull request:
@LeeZion94

STEP 1 묵찌빠 프로젝트 PR 입니다.

Jin(이하 진)이 swift에 익숙하지 않아 easy(이하 이지)가 네비게이터를 맡고 진이 드라이버를 맡아 프로젝트를 진행 했습니다.

과제 요구사항에 맞게 구현을 했지만 부족한 점이 많은 것 같아 피드백 부탁드립니다.

궁금한점

사용자의 입력을 받아 input 변수에 할당 했지만 input은 int 타입이고 readline은 string이라 타입이 달라 할당이 안됩니다. nil 코얼레싱 하였지만 다른 방법이 있는지 궁금합니다.

코드를 좀 더 간결하고 멋지게 만들기 위해서는 어떻게 구조를 설계하는게 좋을까요.

wlsdud-dev committed:
feat : 사용자의 입력을 받아 입렵 값에 따른 분기 생성(#1)

wlsdud-dev committed:
feat : 컴퓨터의 선택 랜덤변수 추가, 종료 명령 추가(#2)
컴퓨터의 가위 바위 보 선택 변수를 생성하고 종료 기능을 추가
fix : input 변수에 예외처리

wlsdud-dev committed:
feat : 사용자 입력에 따른 기능 추가(#3)
사용자 입력에 따른 가위 바위 보 승패 결과를 걔산하고 출력
잘못된 입력 기능 추가
fix : 종료 기능 수정

wlsdud-dev committed:
feat : 상대방의 선택 한 숫자 출력 추가(#4)

LeeZion94 commented:
안녕하세요.
진, 이지의 묵찌빠 프로젝트의 리뷰를 맡게된 리뷰어 시온이라고 합니다. 앞으로 잘부탁드립니다.

궁금한 점으로 언급해 주셨던 부분.
1. nil 코얼레싱 말고도 Optional Type을 바인딩 해줄 수 있는 방법은 여러가지가 있습니다.
Optional Binding 이라는 키워드로 검색을 해보시면 좋을 것 같습니다.

2. 저는 가독성이 좋고 제 3자가 봐도 이해할 수 있도록 책임과 기능이 잘 분리되어 있는 코드가 멋있는 코드라고 생각합니다.
이를 위해서는 먼저 객체지향프로그래밍에 대해서 공부해보시는 것은 어떨까요?
또한 변경에 유연하게 대처할 수 있는 여러가지 설계원칙 중 소개해드리고 싶은 것은 SOLID와 Clean Architecture입니다.
하지만 객체지향프로그래밍을 제외하고 위에서 언급한 부분들은 개발을 진행하시면서 추후에 더 공부를 해보시면 좋을 것 같습니다.

질문드리고 싶은 부분
1. 객체지향프로그래밍이란 무엇인가요?
2. 어떤 코드가 멋있고 간결함이라고 생각하시나요?
3. 간결함과 명료함의 차이는 무엇인가요?
4. 컨벤션을 맞춰서 코드를 작성해야하는 이유는 무엇인가요?

수정해주셨으면 하는 부분
- 객체지향프로그래밍으로의 접근을 위해 struct 혹은 class를 사용하여 객체로 묶고 묶은 객체에 맡게 가위바위보를 진행할 수 있게 기능과 역할을 메서드로 분리해주시면 좋을 것 같아요. 위와 같은 접근이 조금 어렵다면 구두로 같이 논의해봐도 좋아요 언제든 DM 부탁드려요.
- 함수는 한가지 일만할 수 있도록 설계하는 것이 좋아요 Swift API Design Guidlines에 따른 이름짓기와 코딩스타일을 먼저 익히고 코드를 수정해시면 제가 해드릴 수 있는 피드백이 더 많아질 수 있을 것 같아요.
- 현재 승부를 판단하는 로직에서 중복이 너무 많아보여요. 중복을 없이 승부를 판단할 수 있지 않을까요? 단순히 사용자의 입력에 따른 승부를 판단하기 보다는 가위 바위 보간의 승패를 다루는 메서드를 만들면 중복된 코드의 양이 줄 수 있을 것 같아요. 그렇게 접근해볼까요?
- 또한, 전체적으로 요구사항에 맞지 않는 코드 같아요. 그렇게 생각한 이유는 출력의 결과 값이 요구사항과 달랐기 때문이에요. 핵심적인 로직을 구현완료했다고 해서 프로그램의 요구사항을 확인하지 않고 제출하지 않는 것은 좋지 않은 습관같아요. 제출전에는 꼭 요구사항을 확인해보는 습관을 들이면 좋을 것 같아요.

----
- 코드 작성의 이유를 여쭤본 부분은 정말 어떠한 의도로 이 코드가 작성된 것인지 저또한 생각해보기 위함이에요. 그러니 어떠한 의도로 이 코드를 작성해주셨는지 설명해주시면 좀 더 좋은 피드백을 드릴 수 있을 것 같아요.

LeeZion94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -6,5 +6,42 @@
 
 import Foundation
```
Foundation을 import 해주신 이유는 무엇인가요?
반드시 필요한 표현이었나요?

yoruck2 replied:
main.swift 에 기본으로 적혀있어서 그냥 습관적으로 냅두었네요 ㅎㅎ;;
혹시 몰라서 놔둔것도 있지만 Foundation의 제공 기능을 사용하지 않는 저희 코드에는 사실상 없어도 되는 프레임워크 였습니다.

LeeZion94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -6,5 +6,42 @@
 import Foundation
 
-print("Hello, World!")
+while true {
```
while 문 내부의 코드의 길이가 조금은 길어보입니다.
기능과 역할별로 함수로 나누어 보는 것은 어떨까요?
진과 이지는 하나의 블럭이 몇줄이상이면 길다고 생각하시나요?

yoruck2 replied:
Jin : 몇줄인지가 중요하다기 보단 기능 요구사항에 맞게 중복과 불필요한 코드가 없는 것이 짧은 블럭이라고 생각합니다.
Easy : 몇줄부터 긴건지 기준을 나누긴 애매하지만 보통 한 화면을 넘어가는 블럭이 불필요하게 긴 블럭인 경우가 많은 것 같습니다. 하지만 기본적으론 Jin의 의견과 같이 줄 개수에 구애받기 보단 기능구현과 객체지향 프로그래밍에 맞게 중복과 불필요한 코드를 없게 만드는게 더 우선순위 일 것 같습니다.

LeeZion94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -6,5 +6,42 @@
+        } else if comChoice == 2 {
+            print("졌습니다!")
+        }else if comChoice == 3 {
+            print("이겼습니다!")
```
이겼을 때에 대한 요구사항이 출력 및 다시 게임진행이 맞나요?
요구사항을 다시한번 확인해볼까요?

yoruck2 replied:
```swift
menu: while true {
    print("가위(1) 바위(2) 보(3)! <종료 : 0> :")
    var input = readLine()
    var comChoice = Int.random(in: 1...3)
    
switch Int(input ?? "a") {
    case 1:
        if comChoice == 1 {
            print("비겼습니다!")
        } else if comChoice == 2 {
            print("졌습니다!")
            break menu
        } else if comChoice == 3 {
            print("이겼습니다!")
            break menu
        }
        print("상대방의 선택: \(comChoice)")
    case 2:
        if comChoice == 2 {
            print("비겼습니다!")
        } else if comChoice == 3 {
            print("졌습니다!")
            break menu
        } else if comChoice == 1 {
            print("이겼습니다!")
            break menu
        }
        print("상대방의 선택: \(comChoice)")
    case 3:
        if comChoice == 3 {
            print("비겼습니다!")
        } else if comChoice == 1 {
            print("졌습니다!")
            break menu
        } else if comChoice == 2 {
            print("이겼습니다!")
            break menu
        }
        print("상대방의 선택: \(comChoice)")
    case 0:
        print("종료합니다")
        break
    default :
        print("잘못된 입력입니다. 다시 시도해주세요.")
        
    }
    
}

```

yoruck2 replied:
저희 요구사항을 꼼꼼히 읽어보지 않았네요 ..ㅠㅠ 일단 현재 버전을 유지하면서 구현하도록 고쳐보았습니다

LeeZion94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -6,5 +6,42 @@
+    default :
+        print("잘못된 입력입니다. 다시 시도해주세요.")
+    }
+    print("상대방의 선택: \(comChoice)")
```
상대방의 선택을 출력하는 것도 요구사항에 명시된 부분일까요?
명시되어 있지 않은 부분이라면 삭제하는 것이 좋아보입니다.

yoruck2 replied:
버디에게 물어봤는데 결과 요구사항만 충족하면 추가적인 기능은 자유롭게 해도 된다고 하셔서 넣어 보았었습니다.. ㅎㅎ;
삭제해도 무방한 부분입니다. 다음 커밋때 반영 할게요!

LeeZion94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -6,5 +6,42 @@
+while true {
+    print("가위(1) 바위(2) 보(3)! <종료 : 0> :")
+    var input = readLine()
+    var comChoice = Int.random(in: 1...3)
```
input, comChoice는 추후에 변경되는 값이 아니지만 var로 선언된 것으로 보여요,
let과 var의 차이점에 대해서 알아보고 설명해주세요.

또한 컴파일러는 아무런 이유 없이 오류를 발생시키지 않아요~!
컴파일러가 발생시키는 오류는 꼭 오류 메세지를 확인해보고 해결해 보는 습관을 가지면 좋을 것 같습니다. -> 구글링을 통해 해결하는 것도 좋아요!

yoruck2 replied:
xcode가 잡아주니까 일단 혹시몰라 var 로 시작했습니다.  완성 후 경고메세지를 확인했는데 수정하는것을 깜빡했네요 🥲

변수나 상수를 정의 한다는것은 특정 메모리의 주소값을 이름과 연결짓는것을 의미합니다.
이때, swift에서는 값을 변경할 수 있게 할땐 var 키워드로, 변경할 수 없게 하려면 let 키워드로 정의합니다.

앞으로 노란오류도 꼼꼼히 확인 하겠습니다!!

LeeZion94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -6,5 +6,42 @@
+            print("비겼습니다!")
+        } else if comChoice == 2 {
+            print("졌습니다!")
+        }else if comChoice == 3 {
```
위의 else if 와는 다른 컨벤션으로 보이네요?
어떠한 기준으로 이와 같은 코드를 작성해주셨나요. 혹시 의도하신 부분일까요?

yoruck2 replied:
실수 입니다 .. 😅

LeeZion94 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -6,5 +6,42 @@
+            print("비겼습니다!")
+        } else if comChoice == 3 {
+            print("졌습니다!")
+        }else if comChoice == 1 {
```
의도적인 컨벤션이 아니라면, 위쪽에서와의 같은 실수로 보여요.
만약 코드를 복사 붙여넣기 한것이라면 혹시 내가 실수하고 있는 것은 아닌지 생각해보면 좋을 것 같아요.
어떠한 경우에서든 복사붙여넣기가 지향되는 경우는 없다고 생각해요.

yoruck2 replied:
아 그렇군요..! 복붙은 지양하고 하더라도 항상 조심스럽게 접근할게요!

wlsdud-dev commented:
질문해주신 부분에 대해 진과 이지가 고민하여 답변을 적어 보았습니다.

1. 객체지향프로그래밍이란 무엇인가요?
사물, 사람, 개념 등 객체에서 공통된 특성과 행위들을 추출하여 하나의 객체로 만들어 서로 상호작용 하도록 로직을 구현하는 패러다임 입니다.

3. 멋잇고 간결한 코드란?
![image](https://media2.giphy.com/media/PdhafX8jlXSeI/giphy.gif?cid=7941fdc69q1dw3tu6t4j7t5ki9et85sw8hop09tc2a3jni3o&ep=v1_gifs_search&rid=giphy.gif&ct=g)
적절한 추상화를 통해 현실세계의 편리한 기계들의 동작을 반영한 듯한 구조로 누가 보아도 이해가 어렵지 않고 유지보수가 편리한 코드가 멋진 코드 같습니다. (위 시계의 태엽기능 처럼)
그리고 구현해야 할 내용이 방대 하더라도 낭비되거나 중복되는 코드없이 모든 코드가 맞물려가며 의도한 기능을 완벽하게 수행할 때 멋있는 것 같습니다. (오토매틱시계의 컴팩트함 처럼)

4. 간결함과 명료함의 차이 && 4. 컨벤션을 맞춰서 코드를 작성해야하는 이유는 무엇인가요?
코딩에서 간결함이란, 최소한의 코드로 원하는 기능을 구현하는 것이고 
명료함이란 변수와 메서드, 클래스 등의 이름을 해당하는 기능과 목적에 맞게 직관적으로 작성하여 누가 보아도 똑같이 해석될 수 있도록 의미를 전달할 수 있는것을 말합니다. 
이를 위해서 `Swift API Design Guide` 를 준수하여 일관성있는 코딩 컨벤션을 지키고 클린코드를 지향 해야합니다.

yoruck2 commented:
기존코드 작성의도 

사실 저희 둘다 swift 문법과 객체지향에 대해 아직 익숙하지 않아서 일단 요구사항을 구현하는데에 집중했습니다. readline 에서 nil이 나오는 값들은 닐 코얼레싱으로 switch문 에서 자연스럽게 default 케이스로 가도록 예외처리를 해주었습니다.

완성 후 승패와 함께 컴퓨터의 선택도 보여지게 해서 나름의 UI를 구성 했습니다.

코드를 다시 처음부터 짜야할것 같은데, 그 전에 저희가 짠 코드에서 구현되지 않은 사항이 뭔지 알아봤는데 이기고 지는 경우 종료되는것을 놓쳤더라구요. 그래서 break label 로 처리를 해보았고 정상적으로 작동되는것을 확인 했습니다. 하지만 저희의 목표는 객체지향 프로그래밍..!! 이정도만 하고 리펙토링에 들어가려고 합니다

yoruck2 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -6,5 +6,42 @@
+    var input = readLine()
+    var comChoice = Int.random(in: 1...3)
+    
+    switch Int(input ?? "a") {
```
ㅏㅓ

yoruck2 commented:
일단 정의 부터 찾아보았습니다.

### 라이브러리

주로 소프트웨어를 개발할 때 컴퓨터 프로그램이 사용하는 비휘발성 자원의 모임이다. 여기에는 구성 데이터, 문서, 도움말 자료, 메시지 틀, 미리 작성된 코드, 서브루틴(함수), 클래스, 값, 자료형 사양을 포함할 수 있다.

### 프레임워크

애플리케이션 개발 시 필수적인 코드, 알고리즘, DB 커넥션 등의 기능들을 위해 뼈대(구조)를 제공한다. 개발자는 이러한 뼈대 위에서 코드를 작성해 원하는 애플리케이션을 개발할 수 있다.

둘다 좀더 편한 개발을 위해 미리 준비된 코드와 파일등의 모음이라는 것은 알겠는데… 그럼 결정적인 차이점은 무엇일까요?

### 프레임워크 VS 라이브러리

![image](https://github.com/tasty-code/ios-rock-scissor-paper/assets/92171561/53719501-d35a-464f-8cb1-fca21bf5a852)

결정적인 차이점 은 “제어흐름”이 어디에 있는가, 즉 누가 누구를 컨트롤 하는가 입니다.

### 내가 코드를 컨트롤하는가? 누가 정해둔 규칙에 따라 동작하는가?

라이브러리는 개발 중에 내가 원하는 때 호출을 하고, 원하면 얼마든지 대체 가능합니다. 정해진 규칙이 없고 내가 직접 라이브러리를 수정하여 사용할 수도 있습니다.
반면에 프레임워크는 프레임워크가 나를(나의코드) 호출합니다. 프레임워크가 정해둔 규칙에 따라 코딩해야 하고 프레임워크 안에서 코딩을 합니다.

그러나 항상 완벽히 딱 떨어지게 정의 할 수는 없습니다. 관점에 따라 달라지기 때문인데요, 
예를들어 리액트의 경우 기본적으로 라이브러리지만 컴포넌트가 우리코드를 호출하기도 하기 때문에 그 관점에서는 프레임워크적으로 동작한다고 할 수 있습니다.


짧게 정리하면,

**프레임워크:** 개발자가 이미 짜여진 룰을 따라 작업을 하는 것

**라이브러리:** 미리 만들어놓은 필요부분을 사용하는 것

- 공통점: 프로그램을 효율적으로(빨리) 만들기 위해, 다른 사람이 만들어 놓은 코드를 사용하는 것

라고 볼수 있겠습니다!

wlsdud-dev commented:
구조체와 클래스의 차이

## 값타입과 참조타입

- 클래스는 참조타입, 구조체는 값타입 이 제일 큰 차이이다.
- 클래스와 구조체의 같은점
    - 값을 저장하기 위해 프로퍼티를 정의할 수 있다.
    - 기능 실행을 위해 메서드를 정의할 수 있다.
    - 서브 스크립트 구문을 사용하여 값(프로퍼티)에 접근하도록 서브스크립트를 정의할 수 있다.
    - 이니셜라이저를(Initalizers) 정의할 수 있다.
    - 익스텐션(Extension)을 통해 확장할 수 있다.
    - 특정 기능을 실행하기 위해 특정 프로토콜을 준수할 수 있따.
- 클래스와 구조체의 다른점
    - 구조체는 상속할 수 없다.
    - 타입캐스팅은 클래스의 인스턴스에만 허용
    - 디이니셜라이저(Deinitalizers)는 클래스의 인스턴스에만 활용한다.
    - 참조 횟수 계산은 클래스의 인스턴스에만 적용된다.

## 메모리 할당 차이

- 구조체는 stack 영역에 할당되고 클래스는 heap영역에 할당된다
    - 힙은 상대적으로 스택보다 액세스 속도가 느리다.
    - 힙은 참조횟수로 메모리 관리를 해줘야하는 필요가 있다.
    - 스택은 후입선출(LIFO)로 메모리 관리가 쉽다.

wlsdud-dev committed:
refactor : 가위바위보 게임 리팩토링

기존 가위바위보 게임을 리팩토링 하기 위해 enum 추가

wlsdud-dev committed:
feat : 사용자 입력, 컴퓨터 랜덤 선택

사용자 입력값을 받는 기능 추가, 컴퓨터 가위바위보 선택 기능 추가

wlsdud-dev committed:
feat : 사용자, 컴퓨터 가위바위보 판별기능 추가

사용의 선택한 값과 컴퓨터의 랜덤값을 비교해 승패를 판별하고 사용자 입력이 0이라면 게임을 종료한다.

chore : enum의 네이밍을 변경, exit 케이스 추가

wlsdud-dev committed:
feat : 게임 실행 반복 제어문 추가
chore : 오타 수정, 코드 컨벤션

yoruck2 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -2,46 +2,18 @@
+    case draw = "비겼습니다!"
+    case gameOver = "게임 종료"
+    case start = "가위(1) 바위(2) 보(3)! <종료 : 0> :"
+    case error = "잘못된 입력입니다. 다시 시도해주세요."
```
우선 자주 사용되는 값들을 숫자나 문자 리터럴이 아닌 enum의 케이스들로 정의해 사용하도록 구성했습니다.

LeeZion94 replied:
Hand와 ResultMessage를 enum을 사용해 구현해 주신부분을 확인했습니다.
Hand enum의 경우 사용자의 입력을 String으로 입력받아 Int로 만든뒤 Hand enum의 rawValue로 contain을 확인하고 있습니다. 이러한 부분은 의도된 부분이 맞는지 아니면 enum을 사용해보기 위해 오히려 불필요한 로직이 들어간 것이 아닌지 고민해 보시면 좋을 것 같습니다. 코드 구현에 대한 근거만 설명해주셔도 좋아요 ㅎㅎ

또한 ResultMessage의 경우 rawValue로 접근하여 Message 출력하는 것이 주요 목적이므로 case를 사용하는 것보다
enum을 NameSpace로 사용하는 것이 좋아보입니다. static 이라는 키워드에대해서 공부해 보시면 좋을 것 같아요

yoruck2 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,16 +4,25 @@
 //
 
-enum Hand: Int {
+enum Hand: Int, CaseIterable {
```
requestComHand 함수에서 allCases를 활용하기위해 CaseIterable 프로토콜을 채택 했습니다

LeeZion94 replied:
우선 com은 축약표현으로 보입니다. 구두로 논의드렸던대로 가독성을 위해 바꿔보시는게 어떨까요?
또한 CaseIterable이 사용된 부분은 CaseIterable로 하지 않더라도 충분히 구현할 수 있는 표현으로 보입니다. 변경해보시는 건 어떨까요?

yoruck2 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,16 +4,25 @@
+    print("\(Massages.start.rawValue)", terminator: " ")
+    guard let myHandInput = readLine(), let myHandChoice = Int(myHandInput), Hand.scissors.rawValue...Hand.paper.rawValue ~= myHandChoice else { return nil }
+    return myHandChoice
+}
```
사용자의 입력값을 받아 그 정수값을 리턴하는 함수입니다. 
readLine() 과 Int 형변환에서 나오는 nil을 두개의 옵셔널바인딩 조건으로 해결했습니다.

LeeZion94 replied:
구현의 의도는 이해했지만, 깔끔하고 가독성이 좋은 방법은 아니라고 생각됩니다. 사용자의 입력을 받아서 받은 입력을 바탕으로 enum Hand를 할당하는 방법을 사용하면 조금 더 가독성이 좋지 않았을까요? ㅎㅎ

yoruck2 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,16 +4,25 @@
+func requestComHand() -> Int? {
+    guard let comHandChoice = Hand.allCases.randomElement()?.rawValue else { return nil }
+    return comHandChoice
+}
```
컴퓨터의 손패를 결정하는 함수입니다
열거험 Hand의 모든 케이스에서 randomElement() 를 통해 무작위 선택을 하고, 그 케이스의 rawValue를 comHandChoice에 넣고 리턴합니다.

LeeZion94 replied:
이또한 rawValue를 할당하다 보니 rawValue에 접근하기 위한 부분들이 무분별하게 사용되어지고 있는 것 같습니다. 이곳에서도 rawValue보다는 case자체가 할당 되었으면 어땟을까요?

wlsdud-dev committed:
chore : 불필요한 코드 제거

yoruck2 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,7 +4,8 @@
 //
 
-enum Hand: Int, CaseIterable {
+enum UserChoice: Int, CaseIterable {
+    case exit = 0
```
exit 케이스를 추가하면서 열거형 Hand 의 이름이 부적절해서 바꾸었습니다.

LeeZion94 replied:
LGTM

yoruck2 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -19,10 +20,40 @@ enum Massages: String {
+    }
+    if myHandChoice == UserChoice.exit.rawValue {
+        return UserChoice.exit.rawValue
+    }
```
가존 코드는 0을 입력받을 시 nil이 나오게 되어 형변환에 실패 했을때(입력값이 문자열일때)와 같은 결과가 나오는 문제가 있었습니다. 따라서 기존 열거형에 exit = 0 을 추가해 범위 수정을 하고, 0을 받았을 때의 동작도 만들어 주었습니다.

LeeZion94 replied:
이곳에서도case의 비교가 아닌 rawValue 끼리의 비교이기 때문에 조금 어색해 보입니다. 다음 Step에서는 이 부분을 조금 더 신경써주시면 좋을 것 같아요

wlsdud-dev commented:
현재 소스코드는 함수로 만들었는데 클래스나 구조체에 넣어 메서드로 만들어 호출하는게 객체지향적인 프로그래밍일까요?

yoruck2 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -19,10 +20,40 @@ enum Massages: String {
+        print(Massages.gameOver.rawValue)
+        return Massages.lose
+    }
+}
```
requestMyHand 와 requestComHand 함수 에서 리턴한 값을 비교해서 게임의 결과를 판별하는 함수 입니다.

LeeZion94 replied:
judment만으로는 함수의 내용을 설명하기 부족해보입니다. 어떤 걸 판단하는 함수인가요?
또한 2개씩 같이 붙어잇는 print는 1개의 print로 줄여서 출력할 수 있을 것 같습니다. print가 자주 호출되는 것은 퍼포먼스가 좋지 않아서요.

yoruck2 reviewed (commented):
이전에 사용하지 않았던 enum과 함수를 최대한 활용 해서 만들었습니다.
코드를 짜기 전에 사용할 enum 값들과, 한가지 기능을 기준으로 나눈 함수들을 먼저 논의 한 후 체크 해가며 코드를 짰습니다. 덕분에 안정감있는 진행이 가능 했던 것 같습니다.

1. 열거형 값을 rawValue로 가져와 사용시에 코드가 너무 길어지는데 좋은 방법이 있을까요? 아니면 크게 상관이 없을까요? 
2. 전역변수는 쓰지 않는것이 좋나요?
3. 코드 실행부분은 정의부분에 비해 짧고 명료할 수록 좋을까요? 경우에 따라 함수 내에 구현하는 것이 어려워서 실행부에 적기도 하곤 했는데 정의부가 길어지더라도 실행부를 줄이는 것이 캡슐화에 더 맞는걸까요?

LeeZion94 commented:
우선 하나의 Step에 대해서 두번 구현하시느라 정말 고생많으셨어요.
질문을 남겨주셨던 것들이 있어서 답변을 드려요.

1.  이 부분에 관련해서는 코멘트로 남겨놨으니 확인부탁드릴게요
2. 전역변수는 꼭 사용되어야하는 부분에 사용할 수 있겠지만 무분별하게 이유 없이 사용되는 것은 좋지 않습니다. 그 이유는 전역에서 접근하기 때문에 내가 아닌 남에 의해 무분별하게 수정될 수 있고 왜 수정된 것인지도 모르기 때문에 사용하기 위험하기 때문이고 또한 메모리 관리 측면에서도 퍼포먼스가 좋지 않습니다. 전역변수는 메모리의 어디 부분에 저장되는 지 한번 공부해보셔도 좋을 것 같아요.
3. 이 부분은 제가 질문을 제대로 이해하지 못했습니다. 구두 리뷰 시간에 다시 한번 얘기 나눠봐요.
